linera-core/src/client/mod.rs

```diff
@@ -2496,33 +2393,91 @@ where
             }
         }
 
-        // The block we want to propose is either the highest validated, or our pending one.
-        let Some(block) = manager
-            .highest_validated_block()
-            .cloned()
-            .or_else(|| self.state().pending_block().clone())
-        else {
-            return Ok(ClientOutcome::Committed(None)); // Nothing to propose.
+        // Otherwise we have to re-propose the highest validated block, if there is one.
+        let executed_block = if let Some(certificate) = &info.manager.requested_locked {
+            certificate.executed_block().clone()
+        } else {
+            let block = if let Some(proposal) = info
+                .manager
+                .requested_proposed
+                .as_ref()
+                .filter(|proposal| proposal.content.round.is_fast())
+            {
+                // The fast block counts as "locked", too. If there is one, re-propose that.
+                proposal.content.block.clone()
+            } else if let Some(block) = self.state().pending_block() {
+                block.clone() // Otherwise we are free to propose our own pending block.
+            } else {
+                return Ok(ClientOutcome::Committed(None)); // Nothing to do.
+            };
+            self.stage_block_execution(block).await?.0
         };
 
-        let round = match Self::round_for_new_proposal(&info, &identity, Some(&block))? {
-            Either::Left(round) => round,
-            Either::Right(timeout) => return Ok(ClientOutcome::WaitForTimeout(timeout)),
+        let round = match Self::round_for_new_proposal(
+            &info.manager,
+            &identity,
+            Some(&executed_block.block),
+        ) {
+            Ok(round) => round,
+            Err(error) => {
+                let timeout = info.round_timeout().ok_or(error)?;
+                return Ok(ClientOutcome::WaitForTimeout(timeout));
+            }
         };
+
+        let block = executed_block.block.clone();
+        let hashed_value = if round.is_fast() {
+            HashedCertificateValue::new_confirmed(executed_block)
+        } else {
+            HashedCertificateValue::new_validated(executed_block)
+        };
+        // Collect the hashed certificate values required for execution.
+        // Create the final block proposal.
+        let blobs = self.read_local_blobs(block.published_blob_ids()).await?;
```

Comment:
I don't understand: We're only creating the proposal here, based on the block and the current round. And the block is what implies which blobs are being published.

Also, duplicating that information in types that are sent across the network is always dangerous: the receiver would need to check that the information matches. It's more convenient to not have it duplicated in the first place.

Reply:
That would also need discussion with the team. For now, the decision is that blob bodies are never part of blocks, so that blocks and certificates can remain relatively small, even if blobs can be large.

Reply:
Yes, this will be part of the new `Block` structure spec. I think we need to think what do we want from certificates in the first place (how do we expect them to be used and what for).